<!-- TOC -->
* [Exceptions (Исключения)](#exceptions-исключения)
  * [Exception и Error](#exception-и-error)
  * [Checked и Uncheked исключения](#checked-и-uncheked-исключения)
  * [Когда не выполнится блок finally?](#когда-не-выполнится-блок-finally)
  * [Если оператор return содержится и в блоке catch и в finally, какой из них выполнится?](#если-оператор-return-содержится-и-в-блоке-catch-и-в-finally-какой-из-них-выполнится)
  * [Дополнительная информация](#дополнительная-информация)
  * [Полезные ссылки](#полезные-ссылки)
<!-- TOC -->

# Exceptions (Исключения)

![Screenshot](../../../resources/Exceptions.png)

## Exception и Error

Все исключения происходят от `Throwable`. Его потомками являются подклассы `Exception` и `Error`.

**`Exceptions`** являются результатом проблем в программе, которые в принципе решаемые и предсказуемые. Например, произошло деление на ноль в целых числах.

**`Errors`** представляют собой более серьезные проблемы, которые, согласно спецификации Java, не следует пытаться обрабатывать в собственной программе, 
поскольку они связаны с проблемами уровня JVM. Например, исключения такого рода возникают, если закончилась память, доступная виртуальной машине.

## Checked и Uncheked исключения

В Java все исключения делятся на два типа: контролируемые исключения (**checked**) и неконтролируемые исключения (**unchecked**).

**Uncheked** это ошибки (`Errors`) и исключения времени выполнения (`RuntimeExceptions`, потомок класса `Exception`). А также их потомки. 

**Checked** исключения представляют собой ошибки, которые можно и нужно обрабатывать в программе, к этому типу относятся все потомки класса `Exception`, 
кроме `RuntimeException`). **Checked** проверяются на этапе компиляции.

## Когда не выполнится блок finally?

Блок `finally` выполняется почти всегда. Кроме некоторых случаев:

- когда jvm умирает, ей не до `finally`. 
- Если в блоке произошло исключение и нет обработчика, то оставшийся код в блоке `finally` не выполнится.

Пример кода, при котором не выполнится `finally`

```java
try { 
    System.exit(0); 
} catch(Exception e) { 
    e.printStackTrace(); 
} finally { }
```

## Если оператор return содержится и в блоке catch и в finally, какой из них выполнится?

Пример:
```java
public static void main(String[] args) {
    String what = method();
    System.out.println(what);
}

public static String method() {
    try {
        return "SomeString";
    } catch(Exception ex) {
        return "Catch message";
    } finally {
        return "Finally message";
    }
}
    
// Вывод
// Finally message
```

## Дополнительная информация

Общее правило — обрабатывать исключения нужно от «младшего» к старшему.

Блок `finally` будет выполнен, независимо от того, передано исключение или нет. 

Можно не иметь блок `catch` если есть `finally`. Однако, такое не рекомендуется.

## Полезные ссылки

[Собеседование по Java – исключения (exceptions) (вопросы и ответы)](https://javastudy.ru/interview/exceptions/)
